p2p/src/utils/requester.rs

```diff
@@ -0,0 +1,348 @@
+    pub fn skip(&mut self, participant: PublicKey) {
+        self.excluded.insert(participant);
+    }
+
+    /// Remove an outstanding request.
+    fn remove(&mut self, id: ID) -> Option<(PublicKey, SystemTime)> {
+        let (participant, start, deadline) = self.requests.remove(&id)?;
+        let ids = self
+            .deadlines
+            .get_mut(&deadline)
+            .expect("deadline not found");
+        ids.remove(&id);
+        if ids.is_empty() {
+            self.deadlines.remove(&deadline);
+        }
+        Some((participant, start))
+    }
+
+    /// Ask for a participant to handle a request.
+    pub fn request(&mut self) -> Option<(PublicKey, ID)> {
+        // Look for a participant that can handle request
+        for (participant, _) in self.participants.iter() {
+            // Check if me
+            if *participant == self.crypto.public_key() {
+                continue;
+            }
+
+            // Check if excluded
+            if self.excluded.contains(participant) {
+                continue;
+            }
+
+            // Check if rate limit is exceeded (and update rate limiter if not)
+            if self.rate_limiter.check_key(participant).is_err() {
+                continue;
+            }
+
+            // Compute ID
+            //
+            // As long as we don't have u64 requests outstanding, this is ok.
+            let id = self.id;
+            self.id = self.id.wrapping_add(1);
+
+            // Record request issuance time
+            let now = self.runtime.current();
+            let deadline = now.checked_add(self.timeout).expect("time overflowed");
+            self.requests
+                .insert(id, (participant.clone(), now, deadline));
+            self.deadlines.entry(deadline).or_default().insert(id);
+            return Some((participant.clone(), id));
+        }
+        None
+    }
+
+    /// Resolve an outstanding request.
+    pub fn resolved(&mut self, id: ID) {
+        // Remove request
+        let Some((participant, start)) = self.remove(id) else {
+            return;
+        };
+
+        // Get elapsed time
+        let Ok(elapsed) = self.runtime.current().duration_since(start) else {
+            return;
+        };
+        let elapsed = elapsed.as_millis();
+
+        // Calculate new performance using exponential moving average
+        let Some(past) = self.participants.get(&participant) else {
+            return;
+        };
+        let performance = past / 2 + elapsed / 2;
+        self.participants.put(participant, performance);
+    }
+
+    /// Timeout an outstanding request.
+    pub fn timeout(&mut self, id: ID) {
+        let Some((participant, _)) = self.remove(id) else {
+            return;
+        };
+
+        // Calculate new performance using exponential moving average
+        let Some(past) = self.participants.get(&participant) else {
+            return;
+        };
+        let performance = past / 2 + self.timeout.as_millis() / 2;
```

Comment:
`a/2 + b/2` may be wrong compared to `(a+b)/2`, but does avoid overflow.

For example averaging 7 and 3 this way gives 4: `7/2 + 3/2 = 3 + 1 = 4`

This probably doesn't matter much for milliseconds but could also have some sort of `avg()` util function

Reply:
Also, consider having a `learning_rate` in the config between 0 and 1, where:
`performance = past * (1-learning_rate) + timeout * learning_rate;`

Reply:
I avoided the learning_rate approach because I'd have to use floats (which aren't deterministic).

On the point above, this was lazy of me (and had the sum before). I'm going to update it to do a saturating_add before dividing by 2.